lib/src/main/java/tech/relaycorp/relaydroid/persistence/EncryptedDiskPersistence.kt

```diff
@@ -0,0 +1,84 @@
+package tech.relaycorp.relaydroid.persistence
+
+import android.content.Context
+import androidx.security.crypto.EncryptedFile
+import androidx.security.crypto.MasterKey
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import tech.relaycorp.relaydroid.common.Logging.logger
+import java.io.File
+import java.io.IOException
+import java.util.logging.Level
+import kotlin.coroutines.CoroutineContext
+
+class EncryptedDiskPersistence(
+    private val context: Context,
+    private val coroutineContext: CoroutineContext = Dispatchers.IO,
+    private val rootFolder: String = "relaydroid"
+) : Persistence {
+
+    @Suppress("BlockingMethodInNonBlockingContext")
+    override suspend fun set(location: String, data: ByteArray) {
+        withContext(coroutineContext) {
+            delete(location)
+            buildEncryptedFile(location)
+                .openFileOutput()
+                .use { it.write(data) }
+        }
+    }
+
+    @Suppress("BlockingMethodInNonBlockingContext")
+    override suspend fun get(location: String): ByteArray? = withContext(coroutineContext) {
+        try {
+            buildEncryptedFile(location)
+                .openFileInput()
+                .use { it.readBytes() }
+        } catch (exception: IOException) {
```

Comment:
I know this is the pattern we used in the Android Gateway, but in hindsight, I wish we hadn't done that.

If we want to gracefully handle the case where a file does not exist, we should explicitly test that and return `null` when it doesn't exist -- No logging at all, especially no tracebacks that become a distraction in logcat. And if the file does exist but an `IOException` is still thrown for whatever reason (e.g., it isn't readable), it should be propagated -- Otherwise we could be masking actual issues that need to be addressed.

In other words, I'd suggest something like:

```kotlin
if ! fileExists(location) {
    // We could get rid of this "if" if `IOException` can tell us whether the problem
    // was that the file didn't exist. 
    return null
}
try {
    return buildEncryptedFile(location)
                .openFileInput()
                .use { it.readBytes() }
} catch (exception: IOException) {
    throw PersistenceException(exception, "Failed to read $location")
}
```

Where `PersistenceException` is defined in this library (probably with a better name). I'd be tempted to take it a step further and use the `*Error` suffix instead of `*Exception` to signal to the app developer that they _shouldn't_ need to handle it at runtime.

Thoughts?

Reply:
I wanted to gracefully handle it too, but the API they provide is that `.openFileInput()` throws an `IOException` when the file does not exist 😛 

Of course, we can check ourselves first too, but we still need to handle IOException as a file does not exist, since this won't be atomic. Or we can ignore that possibility, and since we checked before, assume the IOException is for some other un-documented exception.

Reply:
Good point re: atomicity.

I don't think `IOException` has a field we can read to tell if the file didn't exist (short of reading the exception message, which would be too brittle), so how about this:

```kotlin
try {
    return buildEncryptedFile(location)
                .openFileInput()
                .use { it.readBytes() }
} catch (exception: IOException) {
    if ! fileExists(location) {
        return null
    }
    throw PersistenceException(exception, "Failed to read $location")
}
```

It's technically not "atomic" yet because the `IOException` could've been caused if the file existed but we didn't have permission to read it, and then something deleted it before we called `fileExists()`... However, I think we can safely ignore that! :smile: